metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py

```diff
@@ -972,49 +1178,52 @@ def get_workunits_internal(self) -> Iterable[MetadataWorkUnit]:
-                workspace_workunits = self.mapper.generate_container_for_workspace(
-                    workspace
-                )
+        allowed_workspaces = self.get_allowed_workspaces()
+        workspaces_len = len(allowed_workspaces)
 
-                for workunit in workspace_workunits:
-                    # Return workunit to Datahub Ingestion framework
-                    yield workunit
-            for dashboard in workspace.dashboards:
-                try:
-                    # Fetch PowerBi users for dashboards
-                    dashboard.users = self.powerbi_client.get_dashboard_users(dashboard)
-                    # Increase dashboard and tiles count in report
-                    self.reporter.report_dashboards_scanned()
-                    self.reporter.report_charts_scanned(count=len(dashboard.tiles))
-                except Exception as e:
-                    message = f"Error ({e}) occurred while loading dashboard {dashboard.displayName}(id={dashboard.id}) tiles."
-
-                    logger.exception(message, e)
-                    self.reporter.report_warning(dashboard.id, message)
-                # Convert PowerBi Dashboard and child entities to Datahub work unit to ingest into Datahub
-                workunits = self.mapper.to_datahub_work_units(dashboard, workspace)
-                for workunit in workunits:
-                    # Return workunit to Datahub Ingestion framework
-                    yield workunit
-
-            for report in workspace.reports:
-                for work_unit in self.mapper.report_to_datahub_work_units(
-                    report, workspace
-                ):
-                    yield work_unit
+        batch_size = 100  # maximum allowed for powerbi scan
+        num_batches = (workspaces_len + batch_size - 1) // batch_size
+        batches = [
+            allowed_workspaces[i * batch_size : (i + 1) * batch_size]
+            for i in range(num_batches)
+        ]
+        for batch_workspaces in batches:
+            for workspace in self.powerbi_client.fill_workspaces(
+                batch_workspaces, self.reporter
+            ):
+                logger.info(f"Processing workspace id: {workspace.id}")
+
+                if self.source_config.modified_since:
+                    # As modified_workspaces is not idempotent, hence we checkpoint for each powerbi workspace
+                    # Because job_id is used as dictionary key, we have to set a new job_id
+                    # Refer to https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/src/datahub/ingestion/source/state/stateful_ingestion_base.py#L390
+                    self.stale_entity_removal_handler.set_job_id(workspace.id)
+                    self.register_stateful_ingestion_usecase_handler(
+                        self.stale_entity_removal_handler
+                    )
+
+                    yield from auto_stale_entity_removal(
```

Comment:
> Is there any reason to skip auto_stale_entity_removal for workspaces return in modified since flow ?

modified_since will only work in this way. where we create checkpoint state for each workspace id